lib/evmone/eof.cpp

```diff
@@ -90,37 +166,285 @@ std::variant<EOFSectionHeaders, EOFValidationError> validate_eof_headers(
+
+                if (jumpdest < 0 || jumpdest >= code_size)
+                    return false;
+
+                rjumpdests.push_back(static_cast<size_t>(jumpdest));
+            }
+
+            const auto rjumpv_imm_size = size_t{1} + count * REL_OFFSET_SIZE;
+            std::fill_n(
+                immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, rjumpv_imm_size, true);
+            i += rjumpv_imm_size;
+            continue;
+        }
+
+        const auto imm_size = instr::traits[op].immediate_size;
+        std::fill_n(immediate_map.begin() + static_cast<ptrdiff_t>(i) + 1, imm_size, true);
+        i += imm_size;
+    }
+
+    // Check relative jump destinations against immediate locations.
+    for (const auto rjumpdest : rjumpdests)
+        if (immediate_map[rjumpdest])
+            return false;
+
+    return true;
+}
+
+std::pair<EOFValidationError, int32_t> validate_max_stack_height(
+    bytes_view code, size_t func_index, const std::vector<EOF1TypeHeader>& funcs_in_outs)
+{
+    assert(code.size() > 0);
+
+    // Changing this type to smaller integer (uint16_t) have in mind that it can overflow.
+    std::vector<int32_t> stack_heights = std::vector<int32_t>(code.size(), -1);
+    std::stack<size_t> worklist;
+
+    int32_t stack_height = 0;
+    stack_heights[0] = funcs_in_outs[func_index].inputs_num;
+    worklist.push(0);
+
+    size_t i = 0;
+    Opcode opcode = {};
+    std::vector<size_t> successors;
+    while (!worklist.empty())
+    {
+        i = worklist.top();
+        opcode = static_cast<Opcode>(code[i]);
+        worklist.pop();
+
+        auto stack_height_required = instr::traits[opcode].stack_height_required;
+        auto stack_height_change = instr::traits[opcode].stack_height_change;
+
+        if (opcode == OP_CALLF)
+        {
+            auto fid_hi = code[i + 1];
+            auto fid_lo = code[i + 2];
+            auto fid = static_cast<uint16_t>((fid_hi << 8) | fid_lo);
+
+            if (fid >= funcs_in_outs.size())
+                return {EOFValidationError::invalid_code_section_index, 0};
+
+            stack_height_required = static_cast<int8_t>(funcs_in_outs[fid].inputs_num);
+            auto d = funcs_in_outs[fid].outputs_num - stack_height_required;
+            stack_height_change = static_cast<int8_t>(d);
+        }
+
+        stack_height = stack_heights[i];
+        assert(stack_height != -1);
+
+        if (stack_height < stack_height_required)
+            return {EOFValidationError::stack_underflow, -1};
+
+        successors.clear();
+
+        // immediate_size for RJUMPV depends on the code. It's calculated below.
+        if (opcode != OP_RJUMP && !instr::traits[opcode].is_terminating && opcode != OP_RJUMPV)
+        {
+            auto next = i + instr::traits[opcode].immediate_size + 1;
+            if (next >= code.size())
+                return {EOFValidationError::no_terminating_instruction, -1};
```

Comment:
Maybe this should be called "out of bounds jump"?

Reply:
There is actually a `truncated_instruction` error code.

Reply:
This checks for final instruction in the code being non-terminating and not RJUMP (and also final RJUMPV is handled separately)